bip352-testsuite/run_bip352_tests.py

```diff
@@ -0,0 +1,227 @@
+            outputs_calculated.append(output.hex())
+            k += 1
+    outputs_expected = [o[0] for o in send_data_expected['outputs']]
+    if outputs_calculated == outputs_expected:
+        print("Sending test \033[0;32mPASSED. ✓\033[0m")
+        passed_send += 1
+    else:
+        print("Sending test \033[0;31mFAILED. ✖\033[0m")
+        print(f"Calculated outputs: {outputs_calculated}")
+        print(f"Expected outputs: {outputs_expected}")
+
+    ###################### receiving ########################
+    assert len(test_vector['receiving']) >= 1
+    for subtest_nr, receive_data in enumerate(test_vector['receiving']):
+        receive_data_given = receive_data['given']
+        receive_data_expected = receive_data['expected']
+
+        input_pub_keys = []
+        outpoints = []
+        for i in receive_data_given['vin']:
+            pub_key = get_pubkey_from_input(bytes.fromhex(i['prevout']['scriptPubKey']['hex']),
+                bytes.fromhex(i['scriptSig']), bytes.fromhex(i['txinwitness']))
+            input_pub_keys.append(pub_key)
+            outpoints.append((i['txid'], i['vout']))
+        receive_outpoint_L = smallest_outpoint(outpoints)
+
+        # test data sanity check: outpoint_L of send and receive has to match
+        assert receive_outpoint_L == outpoint_L
+        # derive tweak_data and shared_secret
+        plain_pubkeys, xonly_pubkeys = [], []
+        for pubkey in input_pub_keys:
+            if len(pubkey) == 32:
+                xonly_pubkeys.append(pubkey)
+            elif len(pubkey) == 33:
+                plain_pubkeys.append(pubkey)
+            else:
+                assert False
+        tweak_data = secp256k1_glue.silentpayments_create_tweak_data(plain_pubkeys, xonly_pubkeys, receive_outpoint_L)
+        scan_privkey = bytes.fromhex(receive_data_given['key_material']['scan_priv_key'])
+        spend_privkey = bytes.fromhex(receive_data_given['key_material']['spend_priv_key'])
+        # spend pubkey is not in the given data of the receiver part, so let's compute it
+        scan_pubkey = secp256k1_glue.pubkey_serialize((secp256k1_glue.pubkey_create(scan_privkey)))
+        spend_pubkey = secp256k1_glue.pubkey_serialize((secp256k1_glue.pubkey_create(spend_privkey)))
+        shared_secret = secp256k1_glue.silentpayments_receive_create_shared_secret(tweak_data, scan_privkey)
+        outputs_pubkeys_expected = [o['pub_key'] for o in receive_data_expected['outputs']]
+        outputs_privkeys_expected = [o['priv_key_tweak'] for o in receive_data_expected['outputs']]
+        outputs_scanned = []
+        outputs_privkeys = []
+
+        # scan through outputs
+        k = 0
+        outputs_to_check = receive_data_given['outputs'].copy()
+        while True:
+            if len(outputs_to_check) == 0:
+                break
+            for output_to_check in outputs_to_check:
+                found_sth = False
+                for label_m in [None] + receive_data_given['labels']:
```

Comment:
in "[DEMO-ONLY] add bip352 test vectors running suite using python and ctypes" (https://github.com/bitcoin-core/secp256k1/pull/1471/commits/3deeebd1b5d7f3684e6425be749c643504f7c471):

This is *kinda* cheating in that you are iterating through all of your labels for every output in the transaction that you're scanning. So lets say I have 100k labels and I'm checking a tx with 100 outputs.. I'd doing `2 * 100k * 100` hashes, along with `100k * 100` EC additions, etc.

Since the receiver already has to generate the label to use it, the idea is they would store the label (something like `{(B_spend + label_tweak * G) : label_tweak}` so that when scanning they can calculate `label = tx_output - P_k` and `label = tx_output_negated - P_k`. This means that regardless of the number of labels used, they only have to do at most two look ups in their label cache when scanning.